### PR TITLE
Adapt Win64 EH to recent LDC EH rewrite and refactor EH into ldc.eh package

### PR DESCRIPTION
**Pull request by kinke**

### CODE REVIEW

**Comment by dnadlinger** [comment]
I'm wondering whether it makes sense to merge the modules, or rather abstract out the common parts (dwarf handling, and maybe also the personality function as a template). Those could live in an `ldc.eh` package.

**Comment by kinke** [comment]
I know, see my remark in the 2nd commit. It's just in an intermediate shape right now, trying to streamline the Win64 code and figuring out differences to find a nice signature for the truly common personality function template.
I've just updated my `win64` branches so that they include this PR, and according to the druntime and Phobos unit tests [_edit:_ and dmd-testsuite], there are no regressions. :) It doesn't include your big EH rewrite yet and still uses the collision landing pads, but it can now be used to start working on that CFA address and make the EH changes Win64-compatible again.

**Comment by dnadlinger** [comment]
> and still uses the collision landing pads

It won't, unless you revert all the cleanup changes.

**Comment by kinke** [comment]
That's what I meant, its state is before your big rewrite (~3-4 days old), and I've reverted your single commit wrt. collision pads before that.

I've just tested with a fully rebased variant, and that one passes all druntime & Phobos unit tests except for `core.thread`, i.e., no regression either and `std.uni-debug` is fixed as expected. So the unit tests most likely don't test any exception chains at all. I'll check out how dmd-testsuite manages another time, need some sleep. _Edit_: yep, there are some new dmd-testsuite issues now.

**Comment by kinke** [comment]
Perfect, this already seems to work with latest `merge-2.067` on Win64 - just 1 regression (dmd-testsuite (release and debug): ICE for `runnable\sdtor.d`, type mismatch for store instruction, but that's most likely not related to the EH changes).
Next: refactoring + splitting up the module.

**Comment by kinke** [comment]
Ready to be reviewed. :P

File: src/ldc/eh/common.d [hunk 1]
```diff
@@ -0,0 +1,650 @@
...
+import core.memory : GC;
+import core.stdc.stdio;
+import core.stdc.stdlib;
+import core.stdc.stdarg;
+
+// D runtime function
+extern(C) int _d_isbaseof(ClassInfo oc, ClassInfo c);
+
+// error and exit
+extern(C) void fatalerror(in char* format, ...)
```
Comment (dnadlinger) [inline src/ldc/eh/common.d:19]: Unrelated, but we should really make this non-extern(C). Thankfully, everybody else seems to properly namespace their libraries, so it hasn't quite collided yet.

**Comment by dnadlinger** [comment]
I've read through the diff, and it should be a good first step, provided that it still passes on Linux.

At some point, we should overhaul the code stylistically and pretty up the DWARF handling, though.

File: src/ldc/eh/common.d [hunk 2]
```diff
@@ -0,0 +1,650 @@
...
+
+
+
+/// This is the implementation of the personality function, which is called by
+/// libunwind twice per frame (search phase, unwind phase).
+///
+/// It is responsible to figure out whether we need to stop unwinding because of
+/// a catch block or if there is a finally block to execute by reading the DWARF
+/// EH tables.
+extern(C) auto eh_personality_common(NativeContext)(ref NativeContext nativeContext)
```
Comment (dnadlinger) [inline src/ldc/eh/common.d:391]: Why extern(C)?

File: src/ldc/eh/common.d [hunk 3]
```diff
@@ -0,0 +1,650 @@
...
+            return _Unwind_GetRegionStart (context);
+      }
+
+        default:
+            fatalerror("Unsupported encoding type to get base from.");
+            assert(0);
+    }
+}
+
+void _d_getLanguageSpecificTables(ubyte* data, ref ubyte* callsite, ref ubyte* action, ref ubyte* classinfo_table, ref ubyte ciEncoding)
```
Comment (dnadlinger) [inline src/ldc/eh/common.d:250]: Can have a prettier name, not extern(C).

File: src/ldc/eh/common.d (hunk 2 above)
Comment (kinke) [inline src/ldc/eh/common.d:391]: It already was `extern(C)`, I've just copy&pasted it.

Comment (dnadlinger) [inline src/ldc/eh/common.d:391]: I know. These were just a couple of things I noticed while reading the diff.

**Pull request merged**